File: backend/models/conversation.py
"""Pydantic models for conversations and individual messages.

A Conversation is a full thread between a user and CaseMate, containing an
ordered list of Messages. Each message is tagged with a role and optional
legal area so the UI can render domain-specific badges and the backend
can filter conversations by topic.
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Literal

from pydantic import BaseModel, Field


class Message(BaseModel):
    """A single message within a conversation.

    Attributes:
        role: Who produced this message — the user, the assistant (CaseMate),
              or an error sentinel used for failed API calls.
        content: The full text content of the message.
        timestamp: When the message was created.
        legal_area: Optional legal domain classification for this message,
                    e.g. 'landlord_tenant' or 'employment_rights'.
    """

    role: Literal["user", "assistant", "error"]
    content: str
    timestamp: datetime = Field(default_factory=lambda: datetime.now(UTC))
    legal_area: str | None = None


class Conversation(BaseModel):
    """A complete conversation thread between a user and CaseMate.

    Attributes:
        id: Unique conversation identifier (UUID string).
        user_id: Supabase auth user ID who owns this conversation.
        messages: Ordered list of messages in the conversation.
        legal_area: Primary legal domain for the overall conversation,
                    determined by the classifier on the first user message.
        created_at: When the conversation was started.
        updated_at: When the last message was added.
    """

    id: str
    user_id: str
    messages: list[Message] = Field(default_factory=list)
    legal_area: str | None = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(UTC))

    def add_message(
        self,
        role: Literal["user", "assistant", "error"],
        content: str,
        legal_area: str | None = None,
    ) -> Message:
        """Append a new message to the conversation and update the timestamp.

        Args:
            role: The message sender role.
            content: The text content of the message.
            legal_area: Optional legal domain tag for this message.

        Returns:
            The newly created Message instance.
        """
        message = Message(role=role, content=content, legal_area=legal_area)
        self.messages.append(message)
        self.updated_at = datetime.now(UTC)
        return message

    def to_anthropic_messages(self) -> list[dict[str, str]]:
        """Convert conversation messages to the Anthropic API message format.

        Filters out error messages since the Anthropic API only accepts
        'user' and 'assistant' roles.

        Returns:
            List of dicts with 'role' and 'content' keys suitable for
            the Anthropic messages API.
        """
        return [
            {"role": msg.role, "content": msg.content}
            for msg in self.messages
            if msg.role in ("user", "assistant")
        ]

File: web/lib/shared-types/conversation.ts
/**
 * A single message in a CaseMate chat conversation.
 *
 * @property role - Who sent the message: "user", "assistant" (CaseMate), or "error"
 * @property content - The message text content
 * @property timestamp - When the message was sent
 * @property legalArea - Optional classified legal domain (e.g., "landlord_tenant")
 */
export interface Message {
  role: "user" | "assistant" | "error";
  content: string;
  timestamp: Date;
  legalArea?: string;
}

/**
 * Response from the CaseMate chat API after processing a legal question.
 *
 * @property conversation_id - ID of the conversation this message belongs to
 * @property answer - CaseMate's personalized legal guidance response
 * @property legal_area - Classified legal domain of the question
 * @property suggested_actions - Actions CaseMate suggests (e.g., "Generate demand letter")
 */
export interface ChatResponse {
  conversation_id: string;
  answer: string;
  legal_area: string;
  suggested_actions: string[];
}

/**
 * Abbreviated conversation data for display in the conversation history sidebar.
 *
 * @property id - Unique conversation identifier
 * @property legal_area - Classified legal domain, or null if not yet classified
 * @property updated_at - ISO timestamp of the most recent message
 * @property preview - Truncated text of the first user message
 * @property message_count - Total messages in the conversation
 */
export interface ConversationSummary {
  id: string;
  legal_area: string | null;
  updated_at: string;
  preview: string;
  message_count: number;
}

/**
 * Full conversation data including all messages, used when loading a conversation.
 *
 * @property id - Unique conversation identifier
 * @property user_id - The user who owns this conversation
 * @property messages - Complete array of messages in chronological order
 * @property legal_area - Overall classified legal domain of the conversation
 * @property created_at - ISO timestamp of conversation creation
 * @property updated_at - ISO timestamp of last activity
 */
export interface ConversationDetail {
  id: string;
  user_id: string;
  messages: {
    role: "user" | "assistant" | "error";
    content: string;
    timestamp: string;
    legal_area: string | null;
  }[];
  legal_area: string | null;
  created_at: string;
  updated_at: string;
}
